### PR TITLE
Refs [TASK][35325] Lấy dữ liệu và hiển thị các thể loại phim

### PR DESCRIPTION
**Pull request by trinq-2471**
## Related Tickets
- [#Task 35325: Lấy dữ liệu và hiển thị các thể loại phim](https://edu-redmine.sun-asterisk.vn/issues/35325)
## WHAT
Screenshot from 2021-04-15 13-19-03

## Review Checklist

Category | View Point | Description | Self review | Reviewer2 (name)
--- | --- | --- | --- | ---
Conventions | Sun* coding conventions | https://github.com/framgia/coding-standards/tree/master/eng/android |<li>- [ ] yes</li>|<li>- [ ] yes</li>  
Conventions | Kotlin coding conventions | https://kotlinlang.org/docs/coding-conventions.html |<li>- [ ] yes</li>|<li>- [ ] yes</li>  
Redmine | Sun* Redmine working process  | https://github.com/framgia/Training-Guideline/blob/master/WorkingProcess/redmine/redmine.md |<li>- [ ] yes</li>|<li>- [ ] yes</li>  

## Notes (Optional)
*(Impacted Areas in Application(List features, api, models or services that this PR will affect))*

*(List gem, library third party add new)*

*(Other notes)*

### CODE REVIEW

File: app/src/main/java/com/sun/moviedb_53/ui/genres/GenresFragment.kt [hunk 1]
```diff
@@ -1,16 +1,94 @@
...
+        genres.positionSelected = position
+        listSelectedGenres.add(genres)
+        adapterGenresSelected.setData(listSelectedGenres)
+        genresPresenter.getGenresMovie(Constant.DEFAULT_PAGE, getQueryGenres())
+    }
+    private val onClickSelectedGenres = fun(positionSelected: Int?, positionRemove: Int) {
+        adapterGenres.unselectedGenres(positionSelected)
+        adapterGenresSelected.removeSelectedGenres(positionRemove)
+        genresPresenter.getGenresMovie(Constant.DEFAULT_PAGE, getQueryGenres())
+    }
```
Comment (phong10101998) [inline app/src/main/java/com/sun/moviedb_53/ui/genres/GenresFragment.kt]:
```suggestion
    private lateinit var genresPresenter: GenresPresenter
    private lateinit var adapterGenres: GenresAdapter
    private lateinit var adapterGenresSelected: GenresSelectedAdapter
    
    private var listSelectedGenres = mutableListOf<Genres>()
    
    private val onClickGenres = fun(genres: Genres, position: Int) {
        adapterGenres.selectedGenres(position)
        genres.positionSelected = position
        listSelectedGenres.add(genres)
        adapterGenresSelected.setData(listSelectedGenres)
        genresPresenter.getGenresMovie(Constant.DEFAULT_PAGE, getQueryGenres())
    }
    private val onClickSelectedGenres = fun(positionSelected: Int?, positionRemove: Int) {
        adapterGenres.unselectedGenres(positionSelected)
        adapterGenresSelected.removeSelectedGenres(positionRemove)
        genresPresenter.getGenresMovie(Constant.DEFAULT_PAGE, getQueryGenres())
    }
```

File: app/src/main/java/com/sun/moviedb_53/data/source/remote/MovieRemoteDataSource.kt [hunk 2]
```diff
@@ -57,8 +58,29 @@ class MovieRemoteDataSource : MovieDataSource.Remote {
...
+        GetJsonFromUrl(listener, KeyEntityType.GENRES_MOVIE_ITEM).execute(stringUrl)
+    }
+
     companion object {
         private const val MOVIE_TYPE = "movie/"
+        private const val GENRES_TYPE = "genre/movie/list?"
+        private const val DISCOVER_TYPE = "discover/movie?"
         private var instance: MovieRemoteDataSource? = null
 
         fun getInstance() = instance ?: MovieRemoteDataSource().also { instance = it }
```
Comment (phong10101998) [inline app/src/main/java/com/sun/moviedb_53/data/source/remote/MovieRemoteDataSource.kt]:
```suggestion
        private const val MOVIE_TYPE = "movie/"
        private const val GENRES_TYPE = "genre/movie/list?"
        private const val DISCOVER_TYPE = "discover/movie?"
        
        private var instance: MovieRemoteDataSource? = null

        fun getInstance() = instance ?: MovieRemoteDataSource().also { instance = it }
```

File: app/src/main/java/com/sun/moviedb_53/ui/genres/GenresFragment.kt [hunk 3]
```diff
@@ -1,16 +1,94 @@
...
+        listGenres[0].isClick = true
+        listGenres[0].positionSelected = 0
+        adapterGenres.setData(listGenres)
+        listSelectedGenres.add(listGenres[0])
+        adapterGenresSelected.setData(listSelectedGenres)
+        genresPresenter.getGenresMovie(Constant.DEFAULT_PAGE, listGenres[0].id.toString())
+    }
+
+    override fun onGetGenresMovieSuccess(listMovieGenres: MutableList<GenresMovie?>) {
+    }
```
Comment (phong10101998) [inline app/src/main/java/com/sun/moviedb_53/ui/genres/GenresFragment.kt:84]:
```suggestion

    override fun onGetGenresMovieSuccess(listMovieGenres: MutableList<GenresMovie?>) { }
```

**Review by phong10101998 — APPROVED** [review]

File: app/src/main/java/com/sun/moviedb_53/ui/genres/GenresFragment.kt [hunk 4]
```diff
@@ -1,16 +1,96 @@
...
+        return query
+    }
+
+    override fun onGetGenresSuccess(listGenres: MutableList<Genres>) {
+        listGenres[0].isClick = true
+        listGenres[0].positionSelected = 0
+        adapterGenres.setData(listGenres)
+        listSelectedGenres.add(listGenres[0])
+        adapterGenresSelected.setData(listSelectedGenres)
+        genresPresenter.getGenresMovie(Constant.DEFAULT_PAGE, listGenres[0].id.toString())
```
Comment (tuanlvt-1117) [inline app/src/main/java/com/sun/moviedb_53/ui/genres/GenresFragment.kt]: check nullable

File: app/src/main/java/com/sun/moviedb_53/ui/genres/GenresFragment.kt [hunk 5]
```diff
@@ -1,16 +1,96 @@
...
+
+    override fun onGetGenresMovieSuccess(listMovieGenres: MutableList<GenresMovie?>) {
+    }
+
+    override fun onError(exception: Exception?) {
+        Toast.makeText(requireContext(), exception?.message, Toast.LENGTH_SHORT).show()
+    }
 
     companion object {
+
```
Comment (tuanlvt-1117) [inline app/src/main/java/com/sun/moviedb_53/ui/genres/GenresFragment.kt]: remove blank

File: app/src/main/java/com/sun/moviedb_53/ui/genres/adapter/GenresAdapter.kt [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+import com.sun.moviedb_53.R
+import com.sun.moviedb_53.data.model.Genres
+import kotlinx.android.synthetic.main.item_genres.view.*
+
+class GenresAdapter(
+    private val context: Context,
+    private var onClickListener: (Genres, Int) -> Unit
+) : RecyclerView.Adapter<GenresAdapter.GenresViewHolder>() {
+
+    private var listGenres = mutableListOf<Genres>()
```
Comment (tuanlvt-1117) [inline app/src/main/java/com/sun/moviedb_53/ui/genres/adapter/GenresAdapter.kt]:
```suggestion
    private var genres = mutableListOf<Genres>()
```

File: app/src/main/java/com/sun/moviedb_53/ui/genres/adapter/GenresAdapter.kt [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+    }
+
+    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): GenresViewHolder {
+        val view = LayoutInflater.from(parent.context).inflate(R.layout.item_genres, parent, false)
+        return GenresViewHolder(view)
+    }
+
+    override fun getItemCount(): Int {
+        return listGenres.size
+    }
```
Comment (tuanlvt-1117) [inline app/src/main/java/com/sun/moviedb_53/ui/genres/adapter/GenresAdapter.kt]:
```suggestion
    override fun getItemCount() = listGenres.size ?: 0
    }
```

File: app/src/main/java/com/sun/moviedb_53/ui/genres/adapter/GenresAdapter.kt [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+
+    override fun onBindViewHolder(holder: GenresViewHolder, position: Int) {
+        listGenres[position].let { holder.bindData(context, it, position, onClickListener) }
```
Comment (tuanlvt-1117) [inline app/src/main/java/com/sun/moviedb_53/ui/genres/adapter/GenresAdapter.kt]:
```suggestion
        holder.bindData(context, listGenres[position], position, onClickListener)
```

File: app/src/main/java/com/sun/moviedb_53/ui/genres/adapter/GenresAdapter.kt [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+    }
+
+    class GenresViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
+
+        fun bindData(
+            context: Context,
+            genres: Genres,
+            position: Int,
+            onClickListener: (Genres, Int) -> Unit
+        ) {
```
Comment (tuanlvt-1117) [inline app/src/main/java/com/sun/moviedb_53/ui/genres/adapter/GenresAdapter.kt]:
```suggestion
    class GenresViewHolder(itemView: View, onClickListener: (Genres, Int) -> Unit) : RecyclerView.ViewHolder(itemView) {

        fun bindData(
            context: Context,
            genres: Genres,
            position: Int) {
```

File: app/src/main/java/com/sun/moviedb_53/ui/genres/adapter/GenresAdapter.kt [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+            itemView.let {
+                it.textGenres.apply {
+                    text = genres.name
+                }
```
Comment (tuanlvt-1117) [inline app/src/main/java/com/sun/moviedb_53/ui/genres/adapter/GenresAdapter.kt]:
```suggestion
                it.textGenres.text = genres.name
```

File: app/src/main/java/com/sun/moviedb_53/data/model/Genres.kt [hunk 11]
```diff
@@ -2,7 +2,9 @@ package com.sun.moviedb_53.data.model
 
 data class Genres(
     val id: Int,
-    val name: String
+    val name: String,
+    var isClick : Boolean = false,
```
Comment (tuanlvt-1117) [inline app/src/main/java/com/sun/moviedb_53/data/model/Genres.kt]:
```suggestion
    var isSelected : Boolean = false,
```

File: app/src/main/java/com/sun/moviedb_53/ui/genres/adapter/GenresAdapter.kt [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+                it.setOnClickListener {
+                    if (!genres.isClick) {
+                        onClickListener(genres, position)
+                    }
```
Comment (tuanlvt-1117) [inline app/src/main/java/com/sun/moviedb_53/ui/genres/adapter/GenresAdapter.kt]:
```suggestion
                    if (!genres.isClick) onClickListener(genres, position)
```

File: app/src/main/java/com/sun/moviedb_53/ui/genres/adapter/GenresSelectedAdapter.kt [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+    private var listGenresSelected = mutableListOf<Genres>()
+
+    fun setData(newListGenresSelected: MutableList<Genres>) {
+        listGenresSelected = newListGenresSelected
+        notifyDataSetChanged()
+    }
+
+    fun removeSelectedGenres(position: Int) {
+        listGenresSelected.removeAt(position)
+        notifyDataSetChanged()
```
Comment (tuanlvt-1117) [inline app/src/main/java/com/sun/moviedb_53/ui/genres/adapter/GenresSelectedAdapter.kt]: notifyItemRemoved(position)
    notifyItemRangeChanged(position, mDataSet.size)

File: app/src/main/java/com/sun/moviedb_53/ui/genres/adapter/GenresSelectedAdapter.kt [hunk 14]
```diff
@@ -0,0 +1,60 @@
...
+
+    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): GenresSelectedViewHolder {
+        val view = LayoutInflater.from(parent.context)
+            .inflate(R.layout.item_genres_selected, parent, false)
+        return GenresSelectedAdapter.GenresSelectedViewHolder(view)
+    }
+
+    override fun getItemCount(): Int {
+        return listGenresSelected.size
+    }
```
Comment (tuanlvt-1117) [inline app/src/main/java/com/sun/moviedb_53/ui/genres/adapter/GenresSelectedAdapter.kt]:
```suggestion
    override fun getItemCount() = listGenresSelected.size ?: 0
```

File: app/src/main/java/com/sun/moviedb_53/ui/genres/adapter/GenresSelectedAdapter.kt [hunk 15]
```diff
@@ -0,0 +1,60 @@
...
+
+    fun removeSelectedGenres(position: Int) {
+        listGenresSelected.removeAt(position)
+        notifyDataSetChanged()
+    }
+
+    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): GenresSelectedViewHolder {
+        val view = LayoutInflater.from(parent.context)
+            .inflate(R.layout.item_genres_selected, parent, false)
+        return GenresSelectedAdapter.GenresSelectedViewHolder(view)
```
Comment (tuanlvt-1117) [inline app/src/main/java/com/sun/moviedb_53/ui/genres/adapter/GenresSelectedAdapter.kt]:
```suggestion
        return GenresSelectedViewHolder(view)
```

File: app/src/main/java/com/sun/moviedb_53/ui/genres/adapter/GenresSelectedAdapter.kt [hunk 16]
```diff
@@ -0,0 +1,60 @@
...
+    }
+
+    class GenresSelectedViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
+
+        fun bindData(
+            listGenres: MutableList<Genres>,
+            position: Int,
+            onClickListener: (Int?, Int) -> Unit
+        ) {
+
```
Comment (tuanlvt-1117) [inline app/src/main/java/com/sun/moviedb_53/ui/genres/adapter/GenresSelectedAdapter.kt]: remove blank

**Review by tuanlvt-1117 — APPROVED** [review]

**Pull request merged**